core/src/com/unciv/logic/multiplayer/chat/ChatWebSocketManager.kt

```diff
@@ -0,0 +1,249 @@
+        get() = URLBuilder(
+            UncivGame.Current.onlineMultiplayer.multiplayerServer.getServerUrl()
+        ).apply {
+            appendPathSegments("chat")
+            protocol = if (protocol.isSecure()) URLProtocol.WSS else URLProtocol.WS
+        }.build()
+
+    /*
+     * Delivery not guaranted & failures are mosly ignored.
+     */
+    fun requestMessageSend(message: Message) {
+        startSocket()
+        Concurrency.run("MultiplayerChatSendMessage") {
+            withTimeoutOrNull(1000) {
+                while (session == null) {
+                    delay(50)
+                    yield()
+                }
+            }
+            session?.runCatching {
+                this.sendSerialized(message)
+            }
+        }
+    }
+
+    fun handleWebSocketThrowables(t: Throwable) {
+        println("ChatError: ${t.message}")
+
+        if (errorReconnectionAttempts == 0)
+            relayGlobalMessage("WebSocket connection closed. Cause: ${t.cause}!")
+
+        if (
+            errorReconnectionAttempts == 0 &&
+            t.message?.contains("401") == true
+        ) {
+            relayGlobalMessage("Authentication issue detected! You have to set a password to use Chat.")
+        }
+
+        restartSocket(true)
+    }
+
+    suspend fun startSession() {
+        try {
+            session?.close()
+            session = client.webSocketSession {
+                url(chatUrl)
+                header(HttpHeaders.Authorization, UncivGame.Current.settings.multiplayer.getAuthHeader())
+            }
+
+            session!!.runCatching {
+                if (isActive) {
+                    if (errorReconnectionAttempts == 0) {
+                        println("ChatLog: Connected to WebSocket.")
+                        relayGlobalMessage("Successfully connected to WebSocket server!")
+                    } else if (errorReconnectionAttempts > 0) {
+                        println("ChatLog: Re-established webSocket connection.")
+                        relayGlobalMessage("Successfully re-established WebSocket connection!")
+                    }
+                }
+
+                val gameIds = ChatStore.getGameIds()
+                    .union(UncivGame.Current.onlineMultiplayer.games.mapNotNull { it.preview?.gameId })
+                this.sendSerialized(Message.Join(gameIds.toList()))
+
+                while (this.isActive) {
+                    val response = receiveDeserialized<Response>()
+                    when (response) {
+                        is Response.Chat -> EventBus.send(
+                            ChatMessageReceived(
+                                response.gameId, response.civName, response.message
+                            )
+                        )
+
+                        is Response.Error -> relayGlobalMessage("Error: ${response.message}", "Server")
+                        is Response.JoinSuccess -> Unit
+                    }
+                    yield()
```

Comment:
yield what?

Reply:
yeild() allows other coroutines to run in the meantime. does this not work that way?

Reply:
resolved by https://github.com/yairm210/Unciv/pull/13599/commits/af50657e05b724848cd01b8650f41327ad91650d

Reply:
Oh, so my `yield` is correct? I saw it in a video and learnt it from there. So, when yair asked me about it, I thought maybe I mistook what I learnt about `yield`. And yes, I know what a `Sequence` builder is. Maybe thats why I thought that I might have confused between that `yield` and something similar.

So, should I reintroduce it or is it ok?

Reply:
Ok, I read the `yeild()` docs and reintroduced it in https://github.com/yairm210/Unciv/pull/13599/commits/1be3151be60853554e1f30e63b4051e03f82f875